### PR TITLE
LAI rewrote helper methods as static functions inside class

### PR DESCRIPTION
**Pull request by Isgotkowitz**
Moved helper functions from outside the LeafAreaIndex class inside and made them `@staticmethod`'s, and changed tests accordingly

Edit: changed senescence flag to get rid of redundancy

### CODE REVIEW

File: SC_redesign/Crop_and_Soil/crop/leaf_area_index.py [hunk 1]
```diff
@@ -94,112 +163,44 @@ def add_leaf_area(self):
         SWAT Reference: 5:2.1.18"""
         self.leaf_area_index = max(0, self.previous_leaf_area_index + self.leaf_area_added)
 
-    def check_for_senescence(self):
+    # TODO CHECKME: having this method with an @property decorator and having the is_in_senescence field seems redundant
+    @property
+    def check_for_senescence(self) -> bool:
```
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/crop/leaf_area_index.py]: How about this?
```suggestion
    def is_in_senescence(self) -> bool:
```

Comment (morrowcj) [reply SC_redesign/Crop_and_Soil/crop/leaf_area_index.py]: Also, consider changing line 48. Is the default starting value `False` actually needed?

File: SC_redesign/Crop_and_Soil/crop/leaf_area_index.py [hunk 2]
```diff
@@ -31,30 +32,80 @@ def __init__(self):
...
+        self._lai_shapes = self.determine_lai_shapes(self.first_heat_fraction_point, self.second_heat_fraction_point,
+                                                     self.first_leaf_fraction_point, self.second_leaf_fraction_point)
+
+        self.optimal_leaf_area_fraction = self.determine_optimal_leaf_area_fraction(self.heat_fraction,
+                                                                                    self._lai_shapes[0],
+                                                                                    self._lai_shapes[1])
+
         self.determine_canopy_height()
-        self.check_for_senescence()
+        self.is_in_senescence = self.check_for_senescence
```
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/crop/leaf_area_index.py]: I believe that `@property` functions don't need to be assigned to variables.
```suggestion
        self.is_in_senescence
```

File: SC_redesign/Crop_and_Soil/tests_SC/test_leaf_area_index.py [hunk 3]
```diff
@@ -273,4 +175,4 @@ def test_determine_leaf_area(heatfrac):
         assert lai.previous_optimal_leaf_area_fraction == optimal_lai
     else:  # senescence
         assert lai.is_in_senescence is True
```
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/tests_SC/test_leaf_area_index.py:203]: might need to change this line (and 166 above) to reflect my suggestions in previous comments.

**Review by morrowcj — CHANGES_REQUESTED** [review]
There are a few small issues with the is_in_senescence property that I think need to be addressed.

File: SC_redesign/Crop_and_Soil/crop/leaf_area_index.py (hunk 2 above)
Comment (morrowcj) [reply SC_redesign/Crop_and_Soil/crop/leaf_area_index.py]: I actually think I was wrong - after looking at the code in its entirety, I think this line can just be deleted. The next line (45) contains the code `if self.is_in_senescence:`, which will automatically check via the `@property` (I think).

File: SC_redesign/Crop_and_Soil/crop/leaf_area_index.py [hunk 4]
```diff
@@ -31,30 +32,80 @@ def __init__(self):
...
             self.add_leaf_area()
         self.shift_leaf_area_time()
 
-    def _determine_lai_shapes(self):
-        """sets lai shape parameters for optimal leaf area fraction calculations"""
-        self._lai_shapes = calc_shape_parameters(self.first_heat_fraction_point, self.second_heat_fraction_point,
-                                                 self.first_leaf_fraction_point, self.second_leaf_fraction_point)
+    @staticmethod
+    def determine_lai_shapes(first_heat_fraction: float, second_heat_fraction: float,
+                             first_leaf_fraction: float, second_leaf_fraction: float) -> List[float]:
```
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/crop/leaf_area_index.py]: Should this be a "private" function (preceded with an underscore)? I don't think LAI shapes will ever be used outside of the class.

File: SC_redesign/Crop_and_Soil/tests_SC/test_leaf_area_index.py (hunk 3 above)
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/tests_SC/test_leaf_area_index.py:203]: Made suggested changes to `is_in_senescence` and all tests passing

File: SC_redesign/Crop_and_Soil/crop/leaf_area_index.py [hunk 5]
```diff
@@ -31,30 +32,80 @@ def __init__(self):
...
+            raise ValueError("first_leaf_fraction must not be greater than 0 or less than 1")
+        if not 0 < second_leaf_fraction < 1:
+            raise ValueError("second_leaf_fraction must not be greater than 0 or less than 1")
+        if first_heat_fraction == second_heat_fraction:
+            # TODO: perhaps a way to handle this instead of throwing an error would be better
+            #   something like: second_heat_fraction += 1e-9
+            raise ValueError("first_heat_fraction cannot be exactly equal to second_heat_fractions")
+
+        # TODO: add error handling (see test_error_calc_shape_log() for conditions that fail)
+
```
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/crop/leaf_area_index.py]: I'd like to update this todo
```suggestion
        # TODO: need to add any of these errors that get thrown when RuFaS runs to the  `OutputManager`. 
        #    This should probably be done in the `grow_canopy()` function 
        #    I'm still unsure how to do this effectively with warnings raised by static functions. - morrowcj
```

File: SC_redesign/Crop_and_Soil/crop/leaf_area_index.py [hunk 6]
```diff
@@ -31,30 +31,79 @@ def __init__(self):
...
-                                                                          self._lai_shapes[0],
-                                                                          self._lai_shapes[1])
+        return [first_shape, second_shape]
+
+    @staticmethod
+    def determine_optimal_leaf_area_fraction(heat_fraction: float, shape1: float, shape2: float) -> float:
+        """calculates leaf area index fraction, from the optimal leaf area development curve, for the initial period of
+        plant growth
+
+        replaces method calc_optimal_leaf_area_fraction
```
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/crop/leaf_area_index.py]: No need for this during the redesign, but I appreciate the thoughtfulness.
```suggestion
```

go ahead and remove similar strings throughout.

File: SC_redesign/Crop_and_Soil/crop/leaf_area_index.py [hunk 7]
```diff
@@ -31,30 +31,79 @@ def __init__(self):
...
-    def _determine_lai_shapes(self):
-        """sets lai shape parameters for optimal leaf area fraction calculations"""
-        self._lai_shapes = calc_shape_parameters(self.first_heat_fraction_point, self.second_heat_fraction_point,
-                                                 self.first_leaf_fraction_point, self.second_leaf_fraction_point)
+    @staticmethod
+    def determine_lai_shapes(first_heat_fraction: float, second_heat_fraction: float,
+                             first_leaf_fraction: float, second_leaf_fraction: float) -> List[float]:
+        """
+        calculates the shape coefficients for optimal LAI formula
+        replaces method calc_shape_parameters
```
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/crop/leaf_area_index.py]:
```suggestion
```

**Review by morrowcj — CHANGES_REQUESTED** [review]
This is a more thorough review, after the example review that we walked through together. I've left a few more in-line comments on the code and have a few more general comments below. Once these issues are addressed, I'll approve the PR. 

## General comments:

* I think that `determine_canopy_height()` should be converted to a static function. `shift_leaf_area_time()` and
`check_previous_leaf_area_values()`can remain as regular methods. 

* Type declarations:  on line 33, `grow_canopy()` needs a function type declaration (i.e., void function: `-> None`). 
Check the remaining functions as well - `shift_leaf_area_time()`, `determine_canopy_height()`, and
`check_previous_leaf_area_values()` all need typing. Caution: they won't all be void functions if you address the
previous bullet first.

* I know this is nit-picky, but can we organize the code so that the static functions are grouped together at the bottom of
the class?

**Comment by morrowcj** [comment]
I've requested @ltp23 to look over this PR, since it implements some comments that you made on the last PR concerning these files.

**Comment by morrowcj** [comment]
Before we merge this in, I'd also like to get @PooyaHekmati's opinion about the pattern that we've employed in this PR.

Edit: To be clear for Pooya's sake, the pattern I'm referring to is the use of static methods like so: 

```python
from math import exp

class SomeClass:
    def __init__(self):
        self.x = 1
        self.y = None

    def main_function(self) -> None:
        """update attributes and apply any logic by applying static methods"""
        self.x = self.change_x(self.x, 5)
        self.y = self.find_y(0.23)

    @staticmethod
    def change_x(x: float, plus: float) -> float:
        """performs a calculation and outputs a value"""
        return x + plus

    @staticmethod
    def find_y(power: float) -> float:
        """performs a different calculation"""
        return exp(power)
```

**Comment by morrowcj** [comment]
To clarify and expand upon: 
>* Type declarations:  on line 33, `grow_canopy()` needs a function type declaration (i.e., void function: `-> None`).
>Check the remaining functions as well - `shift_leaf_area_time()`, `determine_canopy_height()`, and
>`check_previous_leaf_area_values()` all need typing. Caution: they won't all be void functions if you address the
>previous bullet first.

I simply meant that you'll need to make sure that the static function returns a float (i.e., `determine_canopy_height() -> float:`), while the others are void functions (they return nothing). Changing the typing like this should not affect the testing at all, as far as I know. 

As I'm sure you know, Python is a dynamically typed language by default but we want to suggest stricter typing, which is done by adding these type annotations to function declarations:

```python
def void_function(float_arg: float, str_arg: str, int_arg: int, bool_arg: bool, default_float_arg: float = 1.0) -> None:
  return None
```

**Comment by Isgotkowitz** [comment]
> To clarify and expand upon:
> 
> > * Type declarations:  on line 33, `grow_canopy()` needs a function type declaration (i.e., void function: `-> None`).
> >   Check the remaining functions as well - `shift_leaf_area_time()`, `determine_canopy_height()`, and
> >   `check_previous_leaf_area_values()` all need typing. Caution: they won't all be void functions if you address the
> >   previous bullet first.
> 
> I simply meant that you'll need to make sure that the static function returns a float (i.e., `determine_canopy_height() -> float:`), while the others are void functions (they return nothing). Changing the typing like this should not affect the testing at all, as far as I know.
> 
> As I'm sure you know, Python is a dynamically typed language by default but we want to suggest stricter typing, which is done by adding these type annotations to function declarations:
> 
> ```python
> def void_function(float_arg: float, str_arg: str, int_arg: int, bool_arg: bool, default_float_arg: float = 1.0) -> None:
>   return NULL
> ```

Ok, I understand

**Review by morrowcj — APPROVED** [review]
This looks excellent! Great first PR! Thanks for your patience and let me know if you have any feedback about my reviewing process (now or in the future).

**Review by ltp23 — APPROVED** [review]
Overall, it looks very good and clean.

**Pull request merged**